crates/hir_ty/src/method_resolution.rs

```diff
@@ -18,48 +18,69 @@ use crate::{
     db::HirDatabase,
     primitive::{self, FloatTy, IntTy, UintTy},
     utils::all_super_traits,
-    ApplicationTy, Canonical, DebruijnIndex, InEnvironment, Scalar, Substs, TraitEnvironment,
-    TraitRef, Ty, TyKind, TypeCtor, TypeWalk,
+    Canonical, DebruijnIndex, InEnvironment, Scalar, Substs, TraitEnvironment, TraitRef, Ty,
+    TyKind, TypeWalk,
 };
 
 /// This is used as a key for indexing impls.
 #[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
 pub enum TyFingerprint {
-    Apply(TypeCtor),
+    Str,
+    Slice,
+    Array,
+    Never,
+    RawPtr(Mutability),
+    Scalar(Scalar),
+    Adt(AdtId),
     Dyn(TraitId),
+    Tuple { cardinality: u16 },
+    ForeignType(TypeAliasId),
+    FnPtr { num_args: u16, is_varargs: bool },
 }
```

Comment:
I'm not 100% sure these should all be here or if something's missing, would be nice if someone could look over this